src/main/java/net/minecraftforge/registries/NamespacedHolderHelper.java

```diff
@@ -88,7 +89,24 @@ Optional<Holder<T>> getHolder(T value)
         return Optional.ofNullable(this.holders.get(value));
     }
 
-    Holder<T> getOrCreateHolder(ResourceKey<T> key)
+    DataResult<Holder<T>> getOrCreateHolder(ResourceKey<T> key)
+    {
+        Holder.Reference<T> ref = this.holdersByName.get(key.location());
+        if (ref == null)
+        {
+            if (this.holderLookup != null)
+                return DataResult.error("This registry can't create new holders without value");
```

Comment:
Can you add `" (requested key" + key + ")"` to the end of this error message like vanilla?
This additional context would also probably be good in the same error message in `getOrCreateHolderOrThrow`